pyteal/ir/tealblock.py

```diff
@@ -228,5 +228,40 @@ def NormalizeBlocks(cls, start: "TealBlock") -> "TealBlock":
 
         return start
 
+    @classmethod
+    def GetReferencedScratchSlots(cls, start: "TealBlock") -> List["ScratchSlot"]:
+        slots: List[ScratchSlot] = []
+
+        for block in TealBlock.Iterate(start):
+            for op in block.ops:
+                slots += op.getSlots()
+
+        return slots
+
+    @classmethod
+    def MatchScratchSlotReferences(
+        cls, actual: List["ScratchSlot"], expected: List["ScratchSlot"]
+    ) -> bool:
+        if len(actual) != len(expected):
+            return False
+
+        commonSlots = set(actual) & set(expected)
+        mapFromActualToExpected: Dict[ScratchSlot, ScratchSlot] = {
+            slot: slot for slot in commonSlots
+        }
+
+        for actualSlot, expectedSlot in zip(actual, expected):
+            if not actualSlot in mapFromActualToExpected:
+                if expectedSlot in mapFromActualToExpected.values():
+                    # this value was already seen
+                    return False
+                mapFromActualToExpected[actualSlot] = expectedSlot
+                continue
+
+            if mapFromActualToExpected[actualSlot] != expectedSlot:
+                return False
+
+        return True
```

Comment:
These lines of code smells algorithmic, I suppose you are trying to decide a map between `actual` and `expected` satisfying some property (at least it feels acyclic).
My understanding is that, these lines of code ensures that
- cardinality of `actual` >= cardinality of `expected`
- overlapping part of `actual` and `expected` are self-mapping
- allowing more than 1 element map from `actual` to `expected`, but not reversely.

Would be nice to add some more explanation in docstring.

Reply:
I added docs for this, let me know if it makes sense

Reply:
looks good!